### PR TITLE
go_1_11: init at 1.11 and set as default

### PR DESCRIPTION
**Pull request by adisbladis**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

note: I want to set this as default but need to run `nox-review` first.

cc @cstrahan @orivej @velovix @Mic92 

I recently started https://github.com/adisbladis/vgo2nix to use the new modules support with Nix.

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
Does `buildGoPackage` need to change to support the new go modules because it doesn't use GOPATH?

**Comment by CMCDragonkai** [comment]
I noticed that `buildGoPackage` with go overridden to this go version and when running it in a nix-shell to try and build something, it replies with:

```
go: cannot use modules with build cache disabled
```

This is because the `$GOCACHE` is `off` inside the `nix-shell`.

Furthermore this is the `$GOPATH`:

```
/nix/store/wmhb04kda7jqzh4jgn94fva8v8iksci6-go-1.11/share/go:/nix/store/wmhb04kda7jqzh4jgn94fva8v8iksci6-go-1.11/share/go:/nix/store/wmhb04kda7jqzh4jgn94fva8v8iksci6-go-1.11/share/go:/nix/store/wmhb04kda7jqzh4jgn94fva8v8iksci6-go-1.11/share/go:/nix/store/wmhb04kda7jqzh4jgn94fva8v8iksci6-go-1.11/share/go:/nix/store/wmhb04kda7jqzh4jgn94fva8v8iksci6-go-1.11/share/go
```

I'm not sure why it's repeating the same path over and over again.

I think this means `buildGoPackage` for Go modules might have work differently.

**Comment by adisbladis** [comment]
> Does buildGoPackage need to change to support the new go modules because it doesn't use GOPATH?

@CMCDragonkai `buildGoPackage` can be oblivious to go modules. The way `vgo2nix` expressions is exactly the same as `go2nix` and `dep2nix`.

**Comment by CMCDragonkai** [comment]
@adisbladis I discovered this after experimenting with it. The problem is that I wish to shared dependencies between what is loaded by Nix and what can be used by the Go 1.11 compiler when in module mode. That way I'm not downloading the same dependency twice, once during `nix-build`, and another time during `nix-shell` and then using `go build... etc`.

FYI: This sharing of deps between building and development is achieved in the Python Nix.

**Comment by adisbladis** [comment]
I have set go 1.11 as default, both in the `go` and `buildGoPackage` attributes.
`nix-review` passes.

**Comment by adisbladis** [comment]
@GrahamcOfBorg build go_1_11

**Comment by adisbladis** [comment]
Hopefully the last darwin issues are now sorted.

@GrahamcOfBorg build go_1_11

**Comment by adisbladis** [comment]
Darwin issues are resolved. I think this PR should be good to go now.

cc @cstrahan @orivej @velovix @Mic92

File: pkgs/development/tools/goa/default.nix [hunk 1]
```diff
@@ -21,5 +21,6 @@ buildGoPackage rec {
     description = "A framework for building microservices in Go using a unique design-first approach";
     license = licenses.mit;
     maintainers = [ maintainers.rushmorem ];
+    broken = true;
```
Comment (Mic92) [inline pkgs/development/tools/goa/default.nix]: cc @rushmorem

File: pkgs/applications/networking/gopher/gopherclient/default.nix [hunk 2]
```diff
@@ -40,5 +40,6 @@ buildGoPackage rec {
     license = licenses.mit;
     maintainers = with maintainers; [ orivej ];
     platforms = platforms.linux;
+    broken = true;
```
Comment (Mic92) [inline pkgs/applications/networking/gopher/gopherclient/default.nix]: cc @orivej

**Review by Mic92 — COMMENTED** [review]
ping maintainers of broken packages.

**Comment by Mic92** [comment]
@GrahamcOfBorg build docker kubernetes

File: pkgs/development/tools/goa/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/development/tools/goa/default.nix]: Goa seemed to build right now, is a just go110 required? https://hydra.nixos.org/build/80201530

Comment (adisbladis) [reply pkgs/development/tools/goa/default.nix]: goa does indeed build with `buildGo110Package`. I recently had build issues with goa so I mistakenly assumed it was _still_ broken.

**Comment by adisbladis** [comment]
@GrahamcOfBorg test docker

File: pkgs/applications/networking/gopher/gopherclient/default.nix (hunk 2 above)
Comment (orivej) [reply pkgs/applications/networking/gopher/gopherclient/default.nix]: I'll fix it.

**Review by Mic92 — APPROVED** [review]

File: pkgs/development/compilers/go/1.11.nix [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  version = "1.11";
+
+  src = fetchFromGitHub {
+    owner = "golang";
+    repo = "go";
+    rev = "go${version}";
+    sha256 = "1k18d6rkijlgzn1zw4wphzcv6a6w9hb1msgrsh1102jb18644f2q";
+  };
+
+  GOCACHE = "off";
```
Comment (teh) [inline pkgs/development/compilers/go/1.11.nix]: I'm not sure this is working. The derivation is 1.1GB on my machine, and on closer inspection it seems to be mostly this directory:

```
$ du -hcs /nix/store/vk289i3s3ki84681q8pallmgcmsjg708-go-1.11/share/go/pkg/obj/go-build
673M	/nix/store/vk289i3s3ki84681q8pallmgcmsjg708-go-1.11/share/go/pkg/obj/go-build
673M	total
```

and there is a README: 

```
$ cat /nix/store/vk289i3s3ki84681q8pallmgcmsjg708-go-1.11/share/go/pkg/obj/go-build/README
This directory holds cached build artifacts from the Go build system.
Run "go clean -cache" if the directory is getting too large.
See golang.org to learn more about Go.
```

Do we want that directory as part of our go derivation?

**Pull request merged**